source/s3_auto_ranged_put.c

```diff
@@ -447,10 +502,19 @@ static bool s_s3_auto_ranged_put_update(
                 goto has_work_remaining;
             }
 
-            /* There is one more request to send after all of the parts (the complete-multipart-upload) but it can't be
-             * done until all of the parts have been completed.*/
-            if (auto_ranged_put->synced_data.num_parts_completed != auto_ranged_put->synced_data.total_num_parts) {
-                goto has_work_remaining;
+            /* There is one more request to send after all the parts (the complete-multipart-upload) but it can't be
+             * done until all the parts have been completed.
+             * TODO: slightly different handling right now between known and
+             * unknown content length. can this be unified?*/
+            if (auto_ranged_put->has_content_length) {
+                if (auto_ranged_put->synced_data.num_parts_completed != auto_ranged_put->synced_data.total_num_parts) {
+                    goto has_work_remaining;
+                }
+            } else {
+                if ((!auto_ranged_put->prepare_data.is_body_stream_at_end) ||
+                    auto_ranged_put->synced_data.num_parts_completed != auto_ranged_put->synced_data.num_parts_sent) {
+                    goto has_work_remaining;
+                }
             }
```

Comment:
maybe continue using `total_num_parts` but add a bool whether or not total_num_parts is final

```
        /* If content length is not known, this becomes true after we read EOF from the body stream */
        bool total_num_parts_known;
```

I'm not really sure if this simplifies or confuses things, just throwing the idea out there